src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -83,56 +82,157 @@ ShenandoahMmuTracker::~ShenandoahMmuTracker() {
 }
 
 void ShenandoahMmuTracker::record(ShenandoahGeneration* generation) {
-  // This is only called by the control thread.
+  // This is only called by the control thread or the VM thread.
```

Comment:
Would it be worthwhile asserting the calling thread's identity here, just to catch the problem early if someone were to try to do it from a different thread in the future? (although i can't imagine why anyone would.)